group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1043,4 +1230,1358 @@ public void onNewMetadataImage(MetadataImage newImage, MetadataDelta delta) {
+        rescheduleGenericGroupMemberHeartbeat(group, member, genericGroupNewMemberJoinTimeoutMs);
+
+        return maybePrepareRebalanceOrCompleteJoin(group, "Adding new member " + memberId + " with group instance id " +
+            request.groupInstanceId() + "; client reason: " + joinReason);
+    }
+
+    /**
+     * Prepare a rebalance if the group is in a valid state. Otherwise, try
+     * to complete the join phase.
+     *
+     * @param group           The group to rebalance.
+     * @param reason          The reason for the rebalance.
+     *
+     * @return The coordinator result that will be appended to the log.
+     */
+    private CoordinatorResult<Void, Record> maybePrepareRebalanceOrCompleteJoin(
+        GenericGroup group,
+        String reason
+    ) {
+        if (group.canRebalance()) {
+            return prepareRebalance(group, reason);
+        } else {
+            return tryCompleteJoinPhase(group);
+        }
+    }
+
+    /**
+     * Prepare a rebalance.
+     *
+     * @param group           The group to rebalance.
+     * @param reason          The reason for the rebalance.
+     *
+     * @return The coordinator result that will be appended to the log.
+     */
+    CoordinatorResult<Void, Record> prepareRebalance(
+        GenericGroup group,
+        String reason
+    ) {
+        // If any members are awaiting sync, cancel their request and have them rejoin.
+        if (group.isInState(COMPLETING_REBALANCE)) {
+            resetAndPropagateAssignmentWithError(group, Errors.REBALANCE_IN_PROGRESS);
+        }
+
+        // If a sync expiration is pending, cancel it.
+        removeSyncExpiration(group);
+
+        boolean isInitialRebalance = group.isInState(EMPTY);
+        if (isInitialRebalance) {
+            // The group is new. Provide more time for the members to join.
+            int delayMs = genericGroupInitialRebalanceDelayMs;
+            int remainingMs = Math.max(group.rebalanceTimeoutMs() - genericGroupInitialRebalanceDelayMs, 0);
+
+            timer.schedule(
+                joinKey(group.groupId()),
+                delayMs,
+                TimeUnit.MILLISECONDS,
+                () -> tryCompleteInitialRebalanceElseSchedule(group, delayMs, remainingMs)
+            );
+        }
+
+        group.transitionTo(PREPARING_REBALANCE);
+
+        log.info("Preparing to rebalance group {} in state {} with old generation {} ({}-{}) (reason: {})",
+            group.groupId(), group.currentState(), group.generationId(),
+            topicPartition.topic(), topicPartition.partition(), reason);
+
+        return isInitialRebalance ? EMPTY_RESULT : tryCompleteJoinElseSchedule(group);
+    }
+
+    /**
+     * Try to complete the join phase. Otherwise, schedule a new join operation.
+     *
+     * @param group The group.
+     *
+     * @return The coordinator result that will be appended to the log.
+     */
+    private CoordinatorResult<Void, Record> tryCompleteJoinElseSchedule(
```

Comment:
nit: We usually use `maybe`. I understand that this comes from the old purgatories but it may be better to use the correct naming convention.